Add test for `traverseDoc` and `findInDoc`

fisker opened the pull request:
## Description

<!-- Please provide a brief summary of your changes: -->

## Checklist

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

<!-- Please DO NOT remove the playground link -->

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

fisker commented on src/document/utils.js:
```diff
@@ -158,37 +158,34 @@ function mapDoc(doc, cb) {
 
-function findInDoc(doc, fn, defaultValue) {
-  let result = defaultValue;
-  let hasStopped = false;
```
This is useless, `onEnter` returns `false` will stop traverse, must left over during refactoring https://github.com/prettier/prettier/pull/442

fisker replied:
Maybe I'm wrong, adding test.

fisker replied:
return false on `onEnter` doesn't stop traverse. https://github.com/prettier/prettier/pull/13349/commits/e3ff2edc6f778797659dbd8a323a4d26a7be3405

fisker commented on src/document/utils.js:
```diff
@@ -158,37 +158,34 @@ function mapDoc(doc, cb) {
+  let result;
+  traverseDoc(doc, (doc) => {
     const maybeResult = fn(doc);
-    if (maybeResult !== undefined) {
```
I don't think we want use `findInDoc` to do find falsely value.

fisker commented on src/document/utils.js:
```diff
@@ -158,37 +158,34 @@ function mapDoc(doc, cb) {
-function willBreak(doc) {
-  return findInDoc(doc, willBreakFn, false);
+function canBreak(doc) {
+  return Boolean(findInDoc(doc, (doc) => doc.type === DOC_TYPE_LINE));
```
This not seems correct to me, line may live in a unbreakable group. But it's out scope of this PR.

thorn0 commented:
The named functions were used to make profiling more convenient as dev tools show those names in reports. I know we don't profile Prettier much these days, but if we need to do that, we might end up having to return these names

thorn0 commented:
Overall, it's already a very simple function. Also it's exported and may be used by plugins. I'd just remove `hasStopped` and call it a day.

thorn0 commented on src/document/utils.js:
```diff
@@ -161,16 +161,20 @@ function mapDoc(doc, cb) {
   function findInDocOnEnterFn(doc) {
+    if (hasStopped) {
+      // Stop processing rest docs
+      return false;
```
Doesn't `return false` stop the further processing immediately? Is the `hasStopped` variable really needed?

fisker replied:
See https://github.com/prettier/prettier/pull/13349#discussion_r952138588

fisker replied:
I thought it does, but it's not. It will keep processing the siblings.

thorn0 replied:
Ah, okay. I see it's needed.

fisker commented:
Okay, I reverted change, so this PR only going to add some tests.

A minor improvement, the `findInDoc` previously keep running function on the remaining docs, but it should not. I moved `hasStopped` check to top.

thorn0 commented on src/document/utils.js:
```diff
@@ -161,16 +161,20 @@ function mapDoc(doc, cb) {
+
   function findInDocOnEnterFn(doc) {
+    if (hasStopped) {
+      // Stop processing rest docs
```
Yes, the top seems to be a more correct place. Let's just not use the word "stop" in the comment:
```suggestion
      // Skip further processing
```

thorn0 commented:
How about renaming `hasStopped` to `shouldSkipFurtherProcessing` and removing the comment?
